File: modules/mcp/exercises/05_client_session/solution/python/main.py
import json, os, urllib.request

def main() -> None:
    base = os.environ.get("CODAM_LABS_MCP_BASE", "http://127.0.0.1:8765/mcp").rstrip("/")
    req = urllib.request.Request(f"{base}/initialize", data=b"{}", method="POST")
    req.add_header("Content-Type", "application/json")
    with urllib.request.urlopen(req, timeout=10) as r:
        data = json.loads(r.read())
    print(f"SESSION_OK:{data['session_id']}")

if __name__ == "__main__":
    main()

File: modules/mcp/exercises/05_client_session/solution/cpp/main.cpp
#include <httplib.h>
#include <nlohmann/json.hpp>
#include <cstdlib>
#include <iostream>
#include <string>

static std::string mcp_base() {
    const char* env = std::getenv("CODAM_LABS_MCP_BASE");
    return env ? std::string(env) : "http://127.0.0.1:8765/mcp";
}

static std::pair<std::string, std::string> parse_url(const std::string& url) {
    std::string u = url;
    while (!u.empty() && u.back() == '/') u.pop_back();
    size_t scheme = u.find("://");
    size_t start = scheme == std::string::npos ? 0 : scheme + 3;
    size_t path = u.find('/', start);
    if (path == std::string::npos) {
        return {u.substr(start), ""};
    }
    return {u.substr(start, path - start), u.substr(path)};
}

int main() {
    auto [host, prefix] = parse_url(mcp_base());
    httplib::Client cli(host);
    cli.set_connection_timeout(15, 0);
    auto res = cli.Post(prefix + "/initialize", "{}", "application/json");
    if (!res || res->status != 200) return 1;
    auto session = nlohmann::json::parse(res->body).at("session_id").get<std::string>();
    std::cout << "SESSION_OK:" << session << "\n";
    return 0;
}
